enhancements/microshift/advertise-version-details.md

```diff
@@ -0,0 +1,335 @@
+resource in an OCP cluster to discover that it is OCP and what version
+is running. Limiting ourselves to this approach has two
+drawbacks. First, not all console users have cluster admin privileges,
+which means that the console cannot show them the cluster version
+details. Second, the ClusterVersion API is not present in a MicroShift
+deployment, so instead [we created a
+ConfigMap](https://github.com/openshift/microshift/pull/776) in the
+`kube-public` namespace containing some basic version details. That
+means a caller must know two different APIs to check the version and
+product they are communicating with. This enhancement describes how we
+can standardize on one to make it easier for all authenticated users
+to obtain basic product version details for a cluster.
+
+### User Stories
+
+As a console user, I want to see the version of OCP I am using so that
+I can refer to the correct version of the documentation and API
+references.
+
+As a console developer, I want to be able to obtain the version of OCP
+that I am using so that the console app can use different
+implementation choices for behaviors that cannot be discovered by
+probing the Kubernetes API versions supported.
+
+As an authenticated API caller, I want to obtain the product name and
+version of the OpenShift variant I am communicating with so that I can
+use it to make decisions on the client side.
+
+As an application developer, I want to be able to determine the
+Kubernetes distribution the application is being installed into at
+runtime so that I only need to implement one set of installation
+logic.
+
+As an OpenShift user without admin privileges, I want to find
+information about the version of the product I am using in a
+consistent way, regardless of whether I am using standalone OpenShift,
+HyperShift, or MicroShift.
+
+### Goals
+
+* Use a standard Kubernetes type, to keep access easy for arbitrary
+  API clients. This also lets us avoid introducing a new CRD, so we do
+  not increase the overhead for MicroShift.
+* Define a standard location for version information that all
+  OpenShift variants can support.
+
+### Non-Goals
+
+* This change does not introduce a new way for a cluster admin to
+  update the version of a cluster (i.e., to perform an upgrade).
+* This change does not introduce a new way for a client to discover
+  that a cluster is being upgraded.
+* This change does not introduce a new way for a client to discover
+  the cluster topology (single-node, HA, external control-plane, etc.).
+* This change does not introduce a new way for a client to discover
+  other properties of the cluster that might currently only be visible
+  using OpenShift-specific APIs.
+* This change does not introduce a way for a cluster consumer to tell
+  the difference between ROSA, ARO, OSD, HyperShift, or standalone
+  OpenShift clusters because we want those to appear the same from the
+  workload's perspective.
```

Comment:
Console has use cases where we do want to know the variant. Here is one example:

https://bugzilla.redhat.com/show_bug.cgi?id=2106675

We'd also like to make other customizations to the Cluster Settings page for OSD/ROSA/ARO clusters.

Reply:
Does anything in the cluster already know the variant today?

Reply:
I don't believe so :/

@cblecker Do you know?

Reply:
I wouldn't block the enhancement on this, but it would be nice to have.

Reply:
AFAIK there is no structured information to tell the difference between a standalone OCP and OSD/ROSA/ARO. Only a bunch of heuristics that we are evaluating for the purposes of Insights Advisor health checks. None of them seems to be 100% reliable.